src/rvsdg/from_tree.rs

```diff
@@ -0,0 +1,412 @@
+//! Convert tree programs to RVSDGs
+
+use std::{iter, rc::Rc};
+
+use bril_rs::{ConstOps, EffectOps, Literal, ValueOps};
+use tree_in_context::{
+    schema::{BaseType, BinaryOp, Expr, Order, RcExpr, TreeProgram, Type, UnaryOp},
+    typechecker::TypeCache,
+};
+
+use super::{BasicExpr, Operand, RvsdgBody, RvsdgFunction, RvsdgProgram, RvsdgType};
+
+type Operands = Vec<Operand>;
+
+/// State needed to translate a tree program to
+/// an RVSDG.
+struct TreeToRvsdg<'a> {
+    program: &'a TreeProgram,
+    /// A cache of types for every expression in program
+    type_cache: &'a TypeCache,
+    nodes: &'a mut Vec<RvsdgBody>,
+    /// The current state edge, threaded through
+    /// as the last arg to all stateful operations.
+    current_state_edge: Operand,
+    /// The current arguments to the tree program
+    /// as RVSDG operands. (doesn't include state edge)
+    current_args: Vec<Operand>,
+}
+
+pub(crate) fn tree_to_rvsdg(tree: &TreeProgram) -> RvsdgProgram {
+    let mut res = RvsdgProgram { functions: vec![] };
+    for func in &tree.functions {
+        res.functions.push(tree_func_to_rvsdg(func.clone(), tree));
+    }
+    res.functions
+        .push(tree_func_to_rvsdg(tree.entry.clone(), tree));
+    res
+}
+
+fn bril_type_from_type(ty: Type) -> bril_rs::Type {
+    match ty {
+        Type::Base(base_ty) => match base_ty {
+            BaseType::IntT => bril_rs::Type::Int,
+            BaseType::BoolT => bril_rs::Type::Bool,
+        },
+        Type::PointerT(ty) => {
+            let base_ty = bril_type_from_type(Type::Base(ty));
+            bril_rs::Type::Pointer(Box::new(base_ty))
+        }
+        Type::TupleT(_) => panic!("Tuple types not supported in RVSDG"),
+        Type::Unknown => panic!("Unknown type in tree_type_to_rvsdg_types"),
+    }
+}
+
+fn convert_func_output_type(ty: Type) -> Option<bril_rs::Type> {
+    match ty {
+        Type::TupleT(inner) => {
+            assert_eq!(
```

Comment:
nit: assert!(inner.is_empty()) ? (and maybe add `{inner:?}` to the assertion string?